### PR TITLE
ENH: Add lazy copy for swapaxes no op

### PR DESCRIPTION
**Pull request by phofl**
- [ ] xref #49473 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by phofl** [comment]
Not totally sure what to do in the other copy case. This can return views in some cases, but copy is performed on the underlying array

File: pandas/core/generic.py [hunk 1]
```diff
@@ -774,9 +774,9 @@ def swapaxes(
         j = self._get_axis_number(axis2)
 
         if i == j:
-            if copy:
-                return self.copy()
-            return self
+            if not copy and copy is not None and not _using_copy_on_write():
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py:777]:
```suggestion
            if copy is False and not _using_copy_on_write():
```

**Comment by jorisvandenbossche** [comment]
> Not totally sure what to do in the other copy case. This can return views in some cases, but copy is performed on the underlying array

What I think we _ideally_ should do is in case that ``self.values`` is a view on the underlying data (i.e. only if we have a single block I assume) and so the swapped values is also a view of the original data, in that case we should construct a new BlockManager with a ref to the original manager's block. 

This can be done here using some internal APIs, but to do this properly, it might be easier to have this as a method on the Manager.

**Comment by phofl** [comment]
I've used ndarray_to_mgr for now, that's basically what we need here.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -774,15 +777,36 @@ def swapaxes(
...
+            new_mgr = ndarray_to_mgr(
+                new_values,
+                new_axes[0],
+                new_axes[1],
+                dtype=None,
+                copy=False,
+                typ="block",
+            )
+            assert isinstance(new_mgr, BlockManager)
+            assert isinstance(self._mgr, BlockManager)
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py:804]: since this is in the `if` check above, not needed to repeat here?

**Review by jorisvandenbossche — APPROVED** [review]
I still think it's a bit cleaner to move this logic into the internals (we will need to same logic in DataFrame.transpose() as well), but fine for me to keep it like this for now.

File: pandas/core/generic.py (hunk 2 above)
Comment (phofl) [reply pandas/core/generic.py:804]: Thought so as well, but mypy disagrees

**Comment by jorisvandenbossche** [comment]
> we will need to same logic in DataFrame.transpose() as well

Another thought here: now that we are limited to max 2 dimensions in NDFrame, isn't swapaxes just the same as transpose? (for the case i and j are not the same) We could then maybe also consolidate those implementations (eg swapaxes calling transpose when appropriate)

File: pandas/core/generic.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/generic.py:804]: Ah .. When do we use a `cast` and when an `assert` to deal with such issues?

Comment (phofl) [reply pandas/core/generic.py:804]: I prefer the cast here, but not sure if there is a general rule

**Comment by phofl** [comment]
> Another thought here: now that we are limited to max 2 dimensions in NDFrame, isn't swapaxes just the same as transpose? (for the case i and j are not the same) We could then maybe also consolidate those implementations (eg swapaxes calling transpose when appropriate)

Sounds good! But would take a look as follow up, the implementations are quite different and want to understand how this would impact stuff. I think transpose is better, as it keeps dtypes more often if possible

**Comment by phofl** [comment]
Merging for now, happy to address comments in a follow up if there are any

**Pull request merged**